### PR TITLE
Standardize how computes and fixes can produce multiple kinds of output

### PR DESCRIPTION
**Pull request by sjplimp**
**Summary**

Allow computes and fixes to produce mutliple styles (global, peratom, local, pergrid) and kinds (scalar, vector, array) of output data.  A few computes already did this, but the general rule was a single style or output per compute or fix.  This PR changes the documentation to say there are no restrictions.  A few changes to existing commands and the syntax for referencing compute/fix data from variables was needed to standardize the new data production and data access rules.

Specifically this PR does the following:

* updates doc pages for the new generality
*  compute voronoi/atom can now produce both peratom and local data
* compute reduce can now process either peratom or local data, but not both (in one command instance), an "inputs" keyword was added to disambiguate
* equal-style variables can access any scalar from global or peratom computes/fixes, using c_ID versus C_ID to disambiguate
* vector-style variables can only access global vectors from computes/fixes (or use an equal-style variable for any scalar data from a compute/fix)
* atom-style variables can only access peratom data from computes/fixes (or use an equal-style variable for any other scalar data from a compute/fix)

**Related Issue(s)**

**Author(s)**

Steve

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

* The compute voronoi/atom peratom option was dropped; it now always produces peratom data.  The local option is now uses to also generate local data.
* An inputs option was added to compute reduce to operation on peratom (default) or local data.  You must use it to operate on local data
* The rules for equal, vector, atom-style variables changed a bit for how they access output from computes and fixes.  An equal-style variable can access any kind of data, using C_ID, instead of c_ID (similar for fixes) to access per-atom data instead of global data.  Vector and atom-style varaiables can only access global vector and per-atom data directly.  They can use an equal-style variable to access other kinds of data indirectly. 

**Implementation Notes**

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by sjplimp** [comment]
@akohlmey @stanmoore1 Full regression testing (unit, example scripts) should be done on this PR, before merging

**Comment by sjplimp** [comment]
@akohlmey I still need to check the Programmer Guide for doc updates on compute/fix outputs

**Comment by akohlmey** [comment]
@sjplimp found and fixed one bug. Running full test suite now.
Please remember to run "make fix-whitespace" in `src` before committing.

**Comment by sjplimp** [comment]
@akohlmey Do not see any PG changes that are needed

**Comment by akohlmey** [comment]
@sjplimp tests derived from the following example inputs are failing:
- snap/in.snap.compute
```
ERROR: Variable db_2_25: Mismatched compute in variable formula (src/variable.cpp:1550)
```
- snap/in.snap.compute.quadratic
```
ERROR: Variable db_2_100: Mismatched compute in variable formula (src/variable.cpp:1550)
```
- voronoi/in.voronoi
```
ERROR: Variable d5a: Mismatched compute in variable formula (src/variable.cpp:1550)
```
- voronoi/in.voronoi.data
```
ERROR: Illegal compute voronoi/atom command (src/VORONOI/compute_voronoi_atom.cpp:114)
Last command: compute 	v2 all voronoi/atom neighbors yes edge_histo 6 peratom no
```
```

**Comment by sjplimp** [comment]
@akohlmey Fixed 3 of the 4 with new syntax in the input scripts.  However for voronoi/in.voronoi.data there seems to be a code issue in compute_voronoi_atom.cpp which I don't understand.  If I put this printf() as the first line of the constructor:
printf("CR CONSTRUCTOR: UNDECIDED PERATOM LOCAL %d %d %d\n",UNDECIDED,PERATOM,LOCAL);
I get this result:
CR CONSTRUCTOR: UNDECIDED PERATOM LOCAL 0 0 1
which is messing up the logic for peratom vs local below and throwing an error.
How is the enum{} giving those 3 values?
Maybe some collision with PERATOM defined elsewhere?

**Comment by sjplimp** [comment]
ah, now I see that enum is redefined (differently) in the compute_reduce.h file.  Is that the preferred place for it now?

**Comment by sjplimp** [comment]
ok - fixed the 4th input script now

**Comment by akohlmey** [comment]
> ah, now I see that enum is redefined (differently) in the compute_reduce.h file. Is that the preferred place for it now?

If you need constants that are defined consistently across multiple computes, they should be defined in compute.h and the equivalent instances in individual .h or .cpp files should be eliminated. If non-computes need to access them, they can use `Compute::<NAME>`.

File: doc/src/variable.rst [hunk 1]
```diff
@@ -1138,128 +1138,145 @@ only defined if an :doc:`atom_style <atom_style>` is being used that
...
+
+The allowed fix references in variable formulas for equal-, vector-,
+and atom-style variables are listed in the following table:
+
++--------+------------+------------------------------------------+
+| equal  | f_ID       | global scalar                            |
+| equal  | f_ID[I]    | element of global vector                 |
+| equal  | f_ID[I][J] | element of global array                  |
+| equal  | F_ID[I]    | element of per-atom vector (I = atom ID) |
+| equal  | F_ID{i}[J] | element of per-atom array (I = atom ID)  |
```
Comment (akohlmey) [inline doc/src/variable.rst]:
```suggestion
| equal  | F_ID[I][J] | element of per-atom array (I = atom ID)  |
```

File: doc/src/variable.rst [hunk 2]
```diff
@@ -1138,128 +1138,145 @@ only defined if an :doc:`atom_style <atom_style>` is being used that
...
+The various allowed compute references in the variable formulas for
+equal-, vector-, and atom-style variables are listed in the following
+table:
+
++--------+------------+------------------------------------------+
+| equal  | c_ID       | global scalar                            |
+| equal  | c_ID[I]    | element of global vector                 |
+| equal  | c_ID[I][J] | element of global array                  |
+| equal  | C_ID[I]    | element of per-atom vector (I = atom ID) |
+| equal  | C_ID{i}[J] | element of per-atom array (I = atom ID)  |
```
Comment (akohlmey) [inline doc/src/variable.rst]:
```suggestion
| equal  | C_ID[I][J] | element of per-atom array (I = atom ID)  |
```

**Review by akohlmey — CHANGES_REQUESTED** [review]
There seems to be some upper/lower case mixup in the variable docs.

File: doc/src/variable.rst (hunk 2 above)
Comment (sjplimp) [reply doc/src/variable.rst]: good catch - done

**Comment by akohlmey** [comment]
@sjplimp Can you please increment the version in src/version.h from "2 Aug 2023" to "3 Aug 2023".
This is hopefully small enough and only in the develop tree (which has the Development flag set as well) so that it won't make people think there was a new release. I need the difference for the automated testing, so can change the inputs in the regression test tree from, e.g.:
```
variable d5a equal c_v1[${i1}][1]
variable d5b equal c_v1[${i2}][1]
variable d5c equal c_v1[${i1}][2]
variable d5d equal c_v1[${i2}][2]
```
to 
```
if $(version<=20230802) then &
  "variable d5a equal c_v1[${i1}][1]" &
  "variable d5b equal c_v1[${i2}][1]" &
  "variable d5c equal c_v1[${i1}][2]" &
  "variable d5d equal c_v1[${i2}][2]" &
else &
  "variable d5a equal C_v1[${i1}][1]" &
  "variable d5b equal C_v1[${i2}][1]" &
  "variable d5c equal C_v1[${i1}][2]" &
  "variable d5d equal C_v1[${i2}][2]"
```
and thus have the tests working for both, the stable version *and* the development branch.

**Comment by akohlmey** [comment]
@sjplimp voronoi tests still have problems. Please see attached comparing outputs for the "fixed" regression test inputs for the stable version and your pull request.
[log.voronoi.data.new.txt](https://github.com/lammps/lammps/files/12413530/log.voronoi.data.new.txt)
[log.voronoi.data.old.txt](https://github.com/lammps/lammps/files/12413531/log.voronoi.data.old.txt)
[log.voronoi.new.txt](https://github.com/lammps/lammps/files/12413534/log.voronoi.new.txt)
[log.voronoi.old.txt](https://github.com/lammps/lammps/files/12413535/log.voronoi.old.txt)

**Review by akohlmey — APPROVED** [review]

**Comment by akohlmey** [comment]
@sjplimp all checks pass now on the CI test servers and GitHub. @stanmoore1 this is in your hands now.

**Comment by akohlmey** [comment]
@sjplimp can you please draft a little text for the "Backward Compatibility" section of the release notes that summarizes the result of this PR?

**Comment by sjplimp** [comment]
@akohlmey I added the info at the top of this page - can't find the future Release notes page

**Comment by stanmoore1** [comment]
Looking at the code, I don't see anything that needs to be ported Kokkos. I will run this through full Kokkos regression testing though just to be safe.

**Review by stanmoore1 — APPROVED** [review]
Code looks OK to me

**Comment by stanmoore1** [comment]
Regression tests passed, this is ready to merge from my POV.

**Pull request merged**